File: Applications/ZylkerKart_Ecommerce/services/product-service/chaos-sdk/chaos-sdk-core/src/main/java/com/site24x7/labs/chaos/fault/resource/ResourceFaultInjector.java
package com.site24x7.labs.chaos.fault.resource;

import com.site24x7.labs.chaos.config.FaultRuleConfig;
import com.site24x7.labs.chaos.engine.ChaosEngine;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Evaluates resource exhaustion fault rules and injects faults via background threads.
 * <p>
 * Fault types handled:
 * - thread_pool_exhaustion: spawns busy-wait threads to consume thread pool capacity
 * - memory_pressure: allocates large byte arrays to consume heap memory
 * - cpu_burn: spawns threads that perform tight computation loops to burn CPU
 * - gc_pressure: rapidly allocates and discards objects to trigger frequent GC pauses
 * - thread_deadlock: creates two threads that deadlock on ReentrantLocks
 * - disk_fill: writes temp files to consume disk space
 * <p>
 * All resource faults are time-bounded with configurable duration.
 * Background threads are daemon threads so they won't prevent JVM shutdown.
 */
public class ResourceFaultInjector {

    private static final Logger log = LoggerFactory.getLogger(ResourceFaultInjector.class);

    private final ChaosEngine engine;
    private final AtomicBoolean resourceFaultActive = new AtomicBoolean(false);

    public ResourceFaultInjector(ChaosEngine engine) {
        this.engine = engine;
    }

    /**
     * Check for active resource fault rules and execute them.
     * Called periodically by the config watcher or on config update.
     * Only one resource fault can be active at a time to prevent cascading failures.
     */
    public void evaluateAndApply() {
        if (!engine.isEnabled()) {
            return;
        }

        // Prevent overlapping resource faults
        if (resourceFaultActive.get()) {
            log.debug("Resource fault already active, skipping");
            return;
        }

        List<FaultRuleConfig> rules = engine.getActiveRules().stream()
                .filter(r -> isResourceFaultType(r.getFaultType()))
                .filter(engine::shouldFire)
                .toList();

        if (rules.isEmpty()) {
            return;
        }

        // Execute the first matching rule
        FaultRuleConfig rule = rules.get(0);
        switch (rule.getFaultType()) {
            case "thread_pool_exhaustion" -> applyThreadPoolExhaustion(rule);
            case "memory_pressure" -> applyMemoryPressure(rule);
            case "cpu_burn" -> applyCpuBurn(rule);
            case "gc_pressure" -> applyGcPressure(rule);
            case "thread_deadlock" -> applyThreadDeadlock(rule);
            case "disk_fill" -> applyDiskFill(rule);
            default -> log.warn("Unknown resource fault type: {}", rule.getFaultType());
        }
    }

    private boolean isResourceFaultType(String faultType) {
        return "thread_pool_exhaustion".equals(faultType)
                || "memory_pressure".equals(faultType)
                || "cpu_burn".equals(faultType)
                || "gc_pressure".equals(faultType)
                || "thread_deadlock".equals(faultType)
                || "disk_fill".equals(faultType);
    }

    private void applyThreadPoolExhaustion(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        int threadCount = getConfigInt(config, "thread_count", 10);
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        threadCount = Math.min(Math.max(threadCount, 1), 50);
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting thread pool exhaustion: {} threads for {}ms", threadCount, durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final int tc = threadCount;
        final long dm = durationMs;
        Thread orchestrator = new Thread(() -> {
            try {
                CountDownLatch latch = new CountDownLatch(tc);
                List<Thread> threads = new ArrayList<>();
                for (int i = 0; i < tc; i++) {
                    Thread t = new Thread(() -> {
                        try {
                            Thread.sleep(dm);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        } finally {
                            latch.countDown();
                        }
                    }, "chaos-thread-exhaust-" + i);
                    t.setDaemon(true);
                    threads.add(t);
                    t.start();
                }
                latch.await(dm + 5000, TimeUnit.MILLISECONDS);
                log.info("Thread pool exhaustion fault completed");
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                resourceFaultActive.set(false);
            }
        }, "chaos-thread-exhaust-orchestrator");
        orchestrator.setDaemon(true);
        orchestrator.start();
    }

    @SuppressWarnings("unused")
    private void applyMemoryPressure(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        int allocationMb = getConfigInt(config, "allocation_mb", 64);
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        allocationMb = Math.min(Math.max(allocationMb, 1), 512);
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting memory pressure: {}MB for {}ms", allocationMb, durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final int mb = allocationMb;
        final long dm = durationMs;
        Thread memThread = new Thread(() -> {
            byte[][] blocks = null;
            try {
                // Allocate memory in 1MB blocks
                blocks = new byte[mb][];
                for (int i = 0; i < mb; i++) {
                    blocks[i] = new byte[1024 * 1024]; // 1MB each
                    // Touch the memory to ensure it's actually allocated (not just virtual)
                    for (int j = 0; j < blocks[i].length; j += 4096) {
                        blocks[i][j] = (byte) i;
                    }
                }
                log.debug("Memory pressure: allocated {}MB", mb);
                Thread.sleep(dm);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } catch (OutOfMemoryError e) {
                log.warn("Memory pressure: OOM during allocation (allocated partial), holding for duration");
                try {
                    Thread.sleep(dm);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                }
            } finally {
                // Release references so GC can reclaim
                blocks = null;
                System.gc(); // Hint to GC
                resourceFaultActive.set(false);
                log.info("Memory pressure fault completed, memory released");
            }
        }, "chaos-memory-pressure");
        memThread.setDaemon(true);
        memThread.start();
    }

    private void applyCpuBurn(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        int threadCount = getConfigInt(config, "thread_count", 2);
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        threadCount = Math.min(Math.max(threadCount, 1), 8);
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting CPU burn: {} threads for {}ms", threadCount, durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final int tc = threadCount;
        final long dm = durationMs;
        Thread orchestrator = new Thread(() -> {
            try {
                CountDownLatch latch = new CountDownLatch(tc);
                long endTime = System.currentTimeMillis() + dm;
                List<Thread> threads = new ArrayList<>();
                for (int i = 0; i < tc; i++) {
                    Thread t = new Thread(() -> {
                        try {
                            // Tight loop that burns CPU
                            while (System.currentTimeMillis() < endTime && !Thread.currentThread().isInterrupted()) {
                                // Perform computation to burn CPU (not just busy-wait)
                                double x = Math.random();
                                for (int j = 0; j < 1000; j++) {
                                    x = Math.sin(x) * Math.cos(x);
                                }
                            }
                        } finally {
                            latch.countDown();
                        }
                    }, "chaos-cpu-burn-" + i);
                    t.setDaemon(true);
                    threads.add(t);
                    t.start();
                }
                latch.await(dm + 5000, TimeUnit.MILLISECONDS);
                log.info("CPU burn fault completed");
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                resourceFaultActive.set(false);
            }
        }, "chaos-cpu-burn-orchestrator");
        orchestrator.setDaemon(true);
        orchestrator.start();
    }

    @SuppressWarnings("unused")
    private void applyGcPressure(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        int allocRateMbPerSec = getConfigInt(config, "allocation_rate_mb_per_sec", 10);
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        allocRateMbPerSec = Math.min(Math.max(allocRateMbPerSec, 1), 100);
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting GC pressure: {}MB/sec for {}ms", allocRateMbPerSec, durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final int rate = allocRateMbPerSec;
        final long dm = durationMs;
        Thread gcThread = new Thread(() -> {
            try {
                long endTime = System.currentTimeMillis() + dm;
                // Calculate sleep interval to achieve desired rate
                // Allocate 1MB at a time, sleep (1000/rate) ms between allocations
                long sleepMs = Math.max(1, 1000 / rate);
                while (System.currentTimeMillis() < endTime && !Thread.currentThread().isInterrupted()) {
                    // Allocate 1MB — short-lived reference so GC must collect it
                    byte[] garbage = new byte[1024 * 1024];
                    // Touch it to ensure actual allocation
                    garbage[0] = 1;
                    garbage[garbage.length - 1] = 1;
                    // Immediately discard reference (garbage becomes eligible for GC)
                    garbage = null;
                    Thread.sleep(sleepMs);
                }
                log.info("GC pressure fault completed");
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } catch (OutOfMemoryError e) {
                log.warn("GC pressure: OOM during allocation, stopping early");
            } finally {
                resourceFaultActive.set(false);
            }
        }, "chaos-gc-pressure");
        gcThread.setDaemon(true);
        gcThread.start();
    }

    private void applyThreadDeadlock(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting thread deadlock for {}ms", durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final long dm = durationMs;
        final ReentrantLock lockA = new ReentrantLock();
        final ReentrantLock lockB = new ReentrantLock();

        // Thread 1: acquire A then B
        Thread t1 = new Thread(() -> {
            lockA.lock();
            try {
                Thread.sleep(50); // Ensure both threads have their first lock
                lockB.lock(); // Will deadlock here
                try {
                    Thread.sleep(dm);
                } finally {
                    lockB.unlock();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                lockA.unlock();
            }
        }, "chaos-deadlock-1");
        t1.setDaemon(true);

        // Thread 2: acquire B then A (opposite order — guaranteed deadlock)
        Thread t2 = new Thread(() -> {
            lockB.lock();
            try {
                Thread.sleep(50); // Ensure both threads have their first lock
                lockA.lock(); // Will deadlock here
                try {
                    Thread.sleep(dm);
                } finally {
                    lockA.unlock();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                lockB.unlock();
            }
        }, "chaos-deadlock-2");
        t2.setDaemon(true);

        // Watchdog thread: waits for duration, then interrupts both deadlocked threads
        Thread watchdog = new Thread(() -> {
            try {
                t1.start();
                t2.start();
                Thread.sleep(dm);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                t1.interrupt();
                t2.interrupt();
                resourceFaultActive.set(false);
                log.info("Thread deadlock fault completed (threads interrupted)");
            }
        }, "chaos-deadlock-watchdog");
        watchdog.setDaemon(true);
        watchdog.start();
    }

    private void applyDiskFill(FaultRuleConfig rule) {
        Map<String, Object> config = rule.getConfig();
        int allocationMb = getConfigInt(config, "allocation_mb", 64);
        long durationMs = getConfigLong(config, "duration_ms", 30000);

        // Enforce safety limits
        allocationMb = Math.min(Math.max(allocationMb, 1), 512);
        durationMs = Math.min(Math.max(durationMs, 1000), 60000);

        log.info("Injecting disk fill: {}MB for {}ms", allocationMb, durationMs);

        if (!resourceFaultActive.compareAndSet(false, true)) {
            return;
        }

        final int mb = allocationMb;
        final long dm = durationMs;
        Thread diskThread = new Thread(() -> {
            List<File> tempFiles = new ArrayList<>();
            try {
                File tmpDir = new File(System.getProperty("java.io.tmpdir"));
                byte[] block = new byte[1024 * 1024]; // 1MB block
                java.util.Arrays.fill(block, (byte) 0xAA);

                for (int i = 0; i < mb; i++) {
                    File f = new File(tmpDir, "chaos-disk-fill-" + i + ".tmp");
                    f.deleteOnExit(); // Safety net
                    try (FileOutputStream fos = new FileOutputStream(f)) {
                        fos.write(block);
                    }
                    tempFiles.add(f);
                }
                log.debug("Disk fill: wrote {}MB of temp files", mb);

                // Hold the files for the specified duration
                Thread.sleep(dm);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } catch (Exception e) {
                log.warn("Disk fill: error during file writing: {}", e.getMessage());
            } finally {
                // Clean up temp files
                for (File f : tempFiles) {
                    if (f.exists()) {
                        f.delete();
                    }
                }
                resourceFaultActive.set(false);
                log.info("Disk fill fault completed, {} temp files cleaned up", tempFiles.size());
            }
        }, "chaos-disk-fill");
        diskThread.setDaemon(true);
        diskThread.start();
    }

    private int getConfigInt(Map<String, Object> config, String key, int defaultValue) {
        if (config == null) return defaultValue;
        Object val = config.get(key);
        if (val instanceof Number) return ((Number) val).intValue();
        if (val instanceof String) {
            try { return Integer.parseInt((String) val); } catch (NumberFormatException e) { return defaultValue; }
        }
        return defaultValue;
    }

    private long getConfigLong(Map<String, Object> config, String key, long defaultValue) {
        if (config == null) return defaultValue;
        Object val = config.get(key);
        if (val instanceof Number) return ((Number) val).longValue();
        if (val instanceof String) {
            try { return Long.parseLong((String) val); } catch (NumberFormatException e) { return defaultValue; }
        }
        return defaultValue;
    }
}

File: Applications/ZylkerKart_Ecommerce/services/auth-service/chaos-sdk/Fault/ResourceFaultInjector.cs
// Site24x7 Labs Chaos SDK for .NET — Resource Fault Injector
//
// Unlike HTTP/DB/Redis faults (triggered per-request), resource faults are
// triggered when the config changes. The injector registers as a
// config-update listener on the ChaosEngine.
//
// 6 fault types:
//  1. thread_pool_exhaustion — Task.Run flooding the ThreadPool.
//  2. memory_pressure        — Allocate byte[] arrays and hold them.
//  3. cpu_burn               — Thread + SpinWait tight loops.
//  4. gc_pressure            — Rapid short-lived byte[] allocations.
//  5. thread_deadlock        — ManualResetEventSlim.Wait() blocking.
//  6. disk_fill              — File.WriteAllBytes temp files.

namespace Site24x7.Chaos.Fault;

/// <summary>
/// Listens for config changes on the <see cref="ChaosEngine"/> and injects
/// resource exhaustion faults. Only one resource fault is active at a time.
/// </summary>
public sealed class ResourceFaultInjector
{
    private readonly ChaosEngine _engine;
    private volatile bool _active;

    public ResourceFaultInjector(ChaosEngine engine)
    {
        _engine = engine;
    }

    /// <summary>
    /// Register this injector as a config-update listener on the engine.
    /// Call once after the engine is started.
    /// </summary>
    public void Install()
    {
        _engine.AddConfigUpdateListener(EvaluateAndApply);
        Console.WriteLine("[chaos-sdk] ResourceFaultInjector installed");
    }

    /// <summary>
    /// Find the first matching resource fault rule and apply it.
    /// </summary>
    public void EvaluateAndApply()
    {
        if (!_engine.Enabled) return;
        if (_active)
        {
            Console.WriteLine("[chaos-sdk] Resource fault already active, skipping");
            return;
        }

        var rules = _engine.FindMatchingRules("");
        foreach (var rule in rules)
        {
            if (!FaultTypes.ResourceFaultTypes.Contains(rule.FaultType))
                continue;
            if (!_engine.ShouldFire(rule))
                continue;

            switch (rule.FaultType)
            {
                case FaultTypes.ThreadPoolExhaustion:
                    ApplyThreadPoolExhaustion(rule);
                    break;
                case FaultTypes.MemoryPressure:
                    ApplyMemoryPressure(rule);
                    break;
                case FaultTypes.CpuBurn:
                    ApplyCpuBurn(rule);
                    break;
                case FaultTypes.GcPressure:
                    ApplyGcPressure(rule);
                    break;
                case FaultTypes.ThreadDeadlock:
                    ApplyThreadDeadlock(rule);
                    break;
                case FaultTypes.DiskFill:
                    ApplyDiskFill(rule);
                    break;
                default:
                    Console.Error.WriteLine($"[chaos-sdk] Unknown resource fault type: {rule.FaultType}");
                    break;
            }
            return; // Only inject one at a time
        }
    }

    // -----------------------------------------------------------------------
    // 1. Thread pool exhaustion — flood the ThreadPool with blocked tasks
    // -----------------------------------------------------------------------

    private void ApplyThreadPoolExhaustion(FaultRuleConfig rule)
    {
        var threadCount = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "thread_count", 10), 1, 50);
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting thread pool exhaustion: {threadCount} tasks for {durationMs}ms");
        _active = true;

        using var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(durationMs));
        var token = cts.Token;

        for (var i = 0; i < threadCount; i++)
        {
            Task.Run(() =>
            {
                try
                {
                    // Block the ThreadPool thread until cancelled
                    token.WaitHandle.WaitOne();
                }
                catch (ObjectDisposedException)
                {
                    // Expected when CTS is disposed
                }
            });
        }

        // Cleanup task
        Task.Run(async () =>
        {
            await Task.Delay(durationMs).ConfigureAwait(false);
            _active = false;
            Console.WriteLine("[chaos-sdk] Thread pool exhaustion fault completed");
        });
    }

    // -----------------------------------------------------------------------
    // 2. Memory pressure — allocate and hold byte[] arrays
    // -----------------------------------------------------------------------

    private void ApplyMemoryPressure(FaultRuleConfig rule)
    {
        var allocationMb = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "allocation_mb", 64), 1, 512);
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting memory pressure: {allocationMb}MB for {durationMs}ms");
        _active = true;

        var blocks = new List<byte[]>(allocationMb);
        try
        {
            for (var i = 0; i < allocationMb; i++)
            {
                var block = new byte[1024 * 1024]; // 1 MB
                // Touch pages to ensure real allocation
                for (var j = 0; j < block.Length; j += 4096)
                    block[j] = (byte)(i & 0xFF);
                blocks.Add(block);
            }
            Console.WriteLine($"[chaos-sdk] Memory pressure: allocated {allocationMb}MB");
        }
        catch (OutOfMemoryException)
        {
            Console.Error.WriteLine($"[chaos-sdk] Memory pressure: allocation error, holding partial ({blocks.Count}MB)");
        }

        // Hold for duration, then release
        Task.Run(async () =>
        {
            await Task.Delay(durationMs).ConfigureAwait(false);
            blocks.Clear();
            GC.Collect(); // Hint GC
            _active = false;
            Console.WriteLine("[chaos-sdk] Memory pressure fault completed, memory released");
        });
    }

    // -----------------------------------------------------------------------
    // 3. CPU burn — Thread + SpinWait tight loops
    // -----------------------------------------------------------------------

    private void ApplyCpuBurn(FaultRuleConfig rule)
    {
        var threadCount = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "thread_count", 2), 1, 8);
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting CPU burn: {threadCount} threads for {durationMs}ms");
        _active = true;

        var deadline = DateTime.UtcNow.AddMilliseconds(durationMs);
        var completed = 0;

        for (var i = 0; i < threadCount; i++)
        {
            var thread = new Thread(() =>
            {
                var x = 1.1;
                while (DateTime.UtcNow < deadline)
                {
                    // Tight math loop — burns CPU
                    for (var j = 0; j < 10000; j++)
                    {
                        x = Math.Sin(x) * Math.Cos(x) + Math.Sqrt(Math.Abs(x));
                    }
                }
                _ = x; // prevent optimisation

                if (Interlocked.Increment(ref completed) == threadCount)
                {
                    _active = false;
                    Console.WriteLine($"[chaos-sdk] CPU burn fault completed ({threadCount} threads)");
                }
            })
            {
                IsBackground = true,
                Name = $"chaos-cpu-burn-{i}",
            };
            thread.Start();
        }
    }

    // -----------------------------------------------------------------------
    // 4. GC pressure — rapid short-lived allocations
    // -----------------------------------------------------------------------

    private void ApplyGcPressure(FaultRuleConfig rule)
    {
        var allocRateMbPerSec = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "allocation_rate_mb_per_sec", 10), 1, 100);
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting GC pressure: {allocRateMbPerSec}MB/sec for {durationMs}ms");
        _active = true;

        Task.Run(async () =>
        {
            var deadline = DateTime.UtcNow.AddMilliseconds(durationMs);
            var intervalMs = Math.Max(1, 1000 / allocRateMbPerSec);

            while (DateTime.UtcNow < deadline)
            {
                // Allocate 1 MB — immediately discard so GC must collect
                var garbage = new byte[1024 * 1024];
                garbage[0] = 1;
                garbage[^1] = 1;
                _ = garbage; // goes out of scope, eligible for GC

                await Task.Delay(intervalMs).ConfigureAwait(false);
            }

            _active = false;
            Console.WriteLine("[chaos-sdk] GC pressure fault completed");
        });
    }

    // -----------------------------------------------------------------------
    // 5. Thread deadlock — ManualResetEventSlim blocking
    // -----------------------------------------------------------------------

    private void ApplyThreadDeadlock(FaultRuleConfig rule)
    {
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting simulated deadlock for {durationMs}ms");
        _active = true;

        const int deadlockedCount = 4;
        var events = new ManualResetEventSlim[deadlockedCount];

        for (var i = 0; i < deadlockedCount; i++)
        {
            var evt = new ManualResetEventSlim(false);
            events[i] = evt;

            var thread = new Thread(() =>
            {
                // Block until the event is set — simulates deadlock
                evt.Wait();
            })
            {
                IsBackground = true,
                Name = $"chaos-deadlock-{i}",
            };
            thread.Start();
        }

        // Release after duration
        Task.Run(async () =>
        {
            await Task.Delay(durationMs).ConfigureAwait(false);
            foreach (var evt in events)
            {
                evt.Set();
                evt.Dispose();
            }
            _active = false;
            Console.WriteLine("[chaos-sdk] Simulated deadlock fault completed");
        });
    }

    // -----------------------------------------------------------------------
    // 6. Disk fill — write temp files
    // -----------------------------------------------------------------------

    private void ApplyDiskFill(FaultRuleConfig rule)
    {
        var allocationMb = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "allocation_mb", 64), 1, 512);
        var durationMs = ConfigHelpers.Clamp(ConfigHelpers.GetInt(rule, "duration_ms", 30000), 1000, 60000);

        Console.WriteLine($"[chaos-sdk] Injecting disk fill: {allocationMb}MB for {durationMs}ms");
        _active = true;

        var tempDir = Path.GetTempPath();
        var tempFiles = new List<string>(allocationMb);
        var block = new byte[1024 * 1024]; // 1 MB
        Array.Fill(block, (byte)0xAA);

        try
        {
            for (var i = 0; i < allocationMb; i++)
            {
                var filePath = Path.Combine(tempDir, $"chaos-disk-fill-{i}.tmp");
                File.WriteAllBytes(filePath, block);
                tempFiles.Add(filePath);
            }
            Console.WriteLine($"[chaos-sdk] Disk fill: wrote {allocationMb}MB of temp files");
        }
        catch (IOException ex)
        {
            Console.Error.WriteLine($"[chaos-sdk] Disk fill: error during writing: {ex.Message}");
        }

        // Hold for duration, then clean up
        Task.Run(async () =>
        {
            await Task.Delay(durationMs).ConfigureAwait(false);
            foreach (var fp in tempFiles)
            {
                try { File.Delete(fp); }
                catch { /* ignore */ }
            }
            _active = false;
            Console.WriteLine($"[chaos-sdk] Disk fill fault completed, {tempFiles.Count} temp files cleaned up");
        });
    }
}
